components/script/dom/bluetooth.rs

```diff
@@ -94,23 +113,33 @@ impl Bluetooth {
 
         // Note: Steps 6-8 are implemented in
         // components/net/bluetooth_thread.rs in request_device function.
-        let (sender, receiver) = ipc::channel().unwrap();
-        let bt_context = Arc::new(Mutex::new(BluetoothContext {
-            promise: Some(TrustedPromise::new(p.clone())),
-            bluetooth: Trusted::new(self),
-        }));
-        let listener = NetworkListener {
-            context: bt_context,
-            script_chan: self.global().r().networking_task_source(),
-            wrapper: None,
-        };
-        ROUTER.add_route(receiver.to_opaque(), box move |message| {
-            listener.notify_response(message.to().unwrap());
-        });
+        let sender = response_async(p, Trusted::new(self));
         self.get_bluetooth_thread().send(BluetoothMethodMsg::RequestDevice(option, sender)).unwrap();
     }
 }
 
+pub fn response_async<T: AsyncBluetoothListener + Reflectable + 'static>(
+        //chan: Box<ScriptChan + Send>,
+        promise: &Rc<Promise>,
+        receiver: Trusted<T>) -> IpcSender<BluetoothResponseMsg> {
```

Comment:
It would make more sense to take a `&T` argument and store the result of `Trusted::new(receiver)` instead.